bench/bench.ml

```diff
@@ -0,0 +1,480 @@
+
+let value_size = 13
+
+let entry_size = key_size + value_size
+
+let ( // ) = Filename.concat
+
+let random_char () = char_of_int (33 + Random.int 94)
+
+let random_string string_size =
+  String.init string_size (fun _i -> random_char ())
+
+module Context = struct
+  module Key = struct
+    type t = string
+
+    let v () = random_string key_size
+
+    let hash = Hashtbl.hash
+
+    let hash_size = 30
+
+    let encode s = s
+
+    let decode s off = String.sub s off key_size
+
+    let encoded_size = key_size
+
+    let equal = String.equal
+
+    let pp s = Fmt.fmt "%s" s
+  end
+
+  module Value = struct
+    type t = string
+
+    let v () = random_string value_size
+
+    let encode s = s
+
+    let decode s off = String.sub s off value_size
+
+    let encoded_size = value_size
+
+    let pp s = Fmt.fmt "%s" s
+  end
+end
+
+let with_stats f =
+  Stats.reset_stats ();
+  let t0 = Sys.time () in
+  let x = f () in
+  let t1 = Sys.time () -. t0 in
+  let stats = Stats.get () in
+  (x, t1, stats)
+
+module Benchmark = struct
+  type result = {
+    time : float;
+    ops_per_sec : float;
+    mbs_per_sec : float;
+    read_amplification_calls : float;
+    read_amplification_size : float;
+    write_amplification_calls : float;
+    write_amplification_size : float;
+  }
+  [@@deriving yojson]
+
+  let run ~with_metrics ~nb_entries f =
+    let _, time, stats = with_stats (fun () -> f ~with_metrics ()) in
+    let nb_entriesf = Float.of_int nb_entries in
+    let entry_sizef = Float.of_int entry_size in
+    let read_amplification_size =
+      Float.of_int stats.bytes_read /. (entry_sizef *. nb_entriesf)
+    in
+    let read_amplification_calls = Float.of_int stats.nb_reads /. nb_entriesf in
+    let write_amplification_size =
+      Float.of_int stats.bytes_written /. (entry_sizef *. nb_entriesf)
+    in
+    let write_amplification_calls =
+      Float.of_int stats.nb_writes /. nb_entriesf
+    in
+    let ops_per_sec = nb_entriesf /. time in
+    let mbs_per_sec = entry_sizef *. nb_entriesf /. 1_048_576. /. time in
+    {
+      time;
+      ops_per_sec;
+      mbs_per_sec;
+      read_amplification_calls;
+      read_amplification_size;
+      write_amplification_calls;
+      write_amplification_size;
+    }
+
+  let pp_result fmt result =
+    Format.fprintf fmt
+      "Total time: %f@\n\
+       Operations per second: %f@\n\
+       Mbytes per second: %f@\n\
+       Read amplification in syscalls: %f@\n\
+       Read amplification in bytes: %f@\n\
+       Write amplification in syscalls: %f@\n\
+       Write amplification in bytes: %f" result.time result.ops_per_sec
+      result.mbs_per_sec result.read_amplification_calls
+      result.read_amplification_size result.write_amplification_calls
+      result.write_amplification_size
+end
+
+let make_bindings_pool nb_entries =
+  Array.init nb_entries (fun _ ->
+      let k = Context.Key.v () in
+      let v = Context.Value.v () in
+      (k, v))
+
+let bindings_pool = ref [||]
```

Comment:
Why do you have `ref` for these arrays?

Reply:
The goal is to save time on startup so that if you're asking for `bench.exe list` for instance, or `bench.exe --help`, then you get immediate output instead of having to wait for one minute for these to fill.
I agree having a `ref` for this is ugly though, I could probably pass these values instead of having them global.

Reply:
Indeed, I haven't realised that in the previous version, populating the array blocks these commands.